File: src/main/java/miretz/ycloud/views/windows/UploadWindow.java
package miretz.ycloud.views.windows;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import miretz.ycloud.models.Document;
import miretz.ycloud.services.DatabaseService;
import miretz.ycloud.services.DocumentService;

import org.apache.log4j.Logger;

import com.vaadin.server.Page;
import com.vaadin.ui.Alignment;
import com.vaadin.ui.Label;
import com.vaadin.ui.Notification;
import com.vaadin.ui.ProgressBar;
import com.vaadin.ui.TextField;
import com.vaadin.ui.UI;
import com.vaadin.ui.Upload;
import com.vaadin.ui.Upload.Receiver;
import com.vaadin.ui.Upload.StartedEvent;
import com.vaadin.ui.Upload.SucceededEvent;
import com.vaadin.ui.Upload.SucceededListener;
import com.vaadin.ui.VerticalLayout;
import com.vaadin.ui.Window;

public class UploadWindow extends Window {

	private static final long serialVersionUID = 1L;

	private final static Logger logger = Logger.getLogger(UploadWindow.class.getName());

	private final Upload upload;
	private final ProgressBar bar = new ProgressBar();;
	private final Label uploadCaption = new Label("Uploading in progress. Please wait...");
	private final Label commentLabel = new Label("File comment:");
	private final TextField commentField = new TextField();

	public UploadWindow(final DocumentService documentService, final DatabaseService databaseService, final String uploadDir, final Document currentFolder) {

		super("Upload File");
		center();

		// Some basic content for the window
		VerticalLayout content = new VerticalLayout();
		content.setMargin(true);
		setContent(content);
		content.setWidth("310px");

		// Disable the close button
		setClosable(true);
		setResizable(false);
		setModal(true);

		bar.setWidth("250px");
		final VerticalLayout progressLayout = new VerticalLayout();
		progressLayout.addComponent(bar);
		progressLayout.setComponentAlignment(bar, Alignment.MIDDLE_CENTER);
		progressLayout.addComponent(uploadCaption);
		progressLayout.setVisible(false);
		progressLayout.setSpacing(true);
		progressLayout.setWidth("100%");
		content.addComponent(progressLayout);
		content.setComponentAlignment(progressLayout, Alignment.MIDDLE_CENTER);

		class FileUploader implements Receiver, SucceededListener {
			private static final long serialVersionUID = 1L;
			public File file;

			@Override
			public OutputStream receiveUpload(String filename, String mimeType) {
				FileOutputStream fos = null;
				try {
					if (filename == null || filename.isEmpty()) {
						throw new FileNotFoundException("No File Selected!");
					}
					file = new File(uploadDir + filename);
					fos = new FileOutputStream(file);
				} catch (final java.io.FileNotFoundException e) {
					new Notification("Could not open file", e.getMessage(), Notification.Type.ERROR_MESSAGE).show(Page.getCurrent());
					logger.error("Could not open file!", e);
					close();
				}
				return fos;
			}

			@Override
			public void uploadSucceeded(SucceededEvent event) {
				UI.getCurrent().setPollInterval(-1);
				try {
					String creator = String.valueOf(getSession().getAttribute("user"));
					documentService.saveThumbnail(event.getFilename());

					Map<String, String> metadata = new HashMap<>();
					metadata.put("creator", creator);
					metadata.put("comment", commentField.getValue());

					Document document = new Document(UUID.randomUUID().toString(), event.getFilename(), currentFolder.getContentId(), metadata, Document.TYPE_FILE);

					databaseService.addDocument(document);
				} catch (IOException e) {
					logger.error("Failed to create thumbnail!", e);
				} finally {
					Notification.show("Upload completed:", event.getFilename(), Notification.Type.HUMANIZED_MESSAGE);
					close();
				}
			}
		}

		FileUploader receiver = new FileUploader();
		upload = new Upload("Please select your file.", receiver);
		upload.addSucceededListener(receiver);
		upload.setButtonCaption("Upload");
		upload.setWidth("300px");

		// for mobile devices
		// upload.setImmediate(true);

		// progressbar update
		upload.addProgressListener(new Upload.ProgressListener() {
			private static final long serialVersionUID = 1L;

			@Override
			public void updateProgress(long readBytes, long contentLength) {
				bar.setValue(new Float(readBytes / (float) contentLength));
			}
		});
		upload.addStartedListener(new Upload.StartedListener() {
			private static final long serialVersionUID = 1L;

			@Override
			public void uploadStarted(StartedEvent event) {
				UI.getCurrent().setPollInterval(500);
				progressLayout.setVisible(true);
				upload.setVisible(false);
				bar.setValue(0f);
				commentField.setVisible(false);
				commentLabel.setVisible(false);
			}
		});

		commentLabel.setVisible(true);
		commentField.setVisible(true);
		commentField.setWidth("250px");

		content.addComponent(commentLabel);
		content.addComponent(commentField);
		content.addComponent(upload);

	}

}

File: src/main/java/miretz/ycloud/views/windows/UploadWindow.kt
package miretz.ycloud.views.windows

import com.vaadin.data.Property
import com.vaadin.server.Page
import com.vaadin.shared.ui.datefield.Resolution
import com.vaadin.ui.*
import com.vaadin.ui.Upload.*
import miretz.ycloud.models.Document
import miretz.ycloud.models.DocumentMetadata
import miretz.ycloud.services.DatabaseService
import miretz.ycloud.services.DocumentService
import org.apache.log4j.Logger
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStream
import java.nio.file.Paths
import java.util.*

class UploadWindow(documentService: DocumentService, databaseService: DatabaseService, uploadDir: String, currentFolder: Document) : Window("Upload File") {

    private val upload: Upload
    private val bar = ProgressBar()
    private val uploadCaption = Label("Uploading in progress. Please wait...")
    private val filenameLabel = Label("File Name:")
    private val filenameField = TextField()
    private val commentLabel = Label("File comment:")
    private val commentField = TextField()
    private val uploadButton = Button("Upload file")
    private val retentionCheckbox = CheckBox("Temporary file");
    private val retentionDateField = DateField();
    private val retentionDateLabel = Label("Delete file on:")

    init {
        center()

        // Some basic content for the window
        val content = VerticalLayout()
        content.setMargin(true)
        content.isSpacing = true
        setContent(content)
        content.setWidth("310px")

        isClosable = true
        isResizable = false
        isModal = true

        bar.setWidth("250px")
        val progressLayout = VerticalLayout()
        progressLayout.addComponent(bar)
        progressLayout.setComponentAlignment(bar, Alignment.MIDDLE_CENTER)
        progressLayout.addComponent(uploadCaption)
        progressLayout.isVisible = false
        progressLayout.isSpacing = true
        progressLayout.setWidth("100%")
        content.addComponent(progressLayout)
        content.setComponentAlignment(progressLayout, Alignment.MIDDLE_CENTER)

        class FileUploader : Receiver, SucceededListener {
            var file: File? = null
            var uid: String = ""

            override fun receiveUpload(filenameV: String?, mimeType: String): OutputStream {
                try {
                    uid = UUID.randomUUID().toString()
                    val fixedDir = if (uploadDir.endsWith(File.separator)) uploadDir else uploadDir + File.separator
                    file = File(fixedDir + uid)
                } catch (e: java.io.FileNotFoundException) {
                    Notification("Could not open file", e.message, Notification.Type.ERROR_MESSAGE).show(Page.getCurrent())
                    logger.error("Could not open file!", e)
                    close()
                }

                return FileOutputStream(file)
            }

            override fun uploadSucceeded(event: SucceededEvent) {
                UI.getCurrent().pollInterval = -1

                val creator = (session.getAttribute("user")) as String

                val metadata = HashMap<String, String>()
                metadata.put(DocumentMetadata.CREATOR.toString(), creator)
                metadata.put(DocumentMetadata.COMMENT.toString(), commentField.value)

                val retentionDate = if (retentionCheckbox.value!!) {
                    retentionDateField.value
                } else {
                    null
                }

                val document = Document(uid, filenameField.value, currentFolder.contentId, metadata, Document.TYPE_FILE, retentionDate)

                documentService.saveThumbnail(document)

                databaseService.addDocument(document)

                Notification.show("Upload completed:", event.filename, Notification.Type.HUMANIZED_MESSAGE)
                close()
            }
        }

        val receiver = FileUploader()
        upload = Upload("Please select your file:", receiver)
        upload.addSucceededListener(receiver)
        upload.buttonCaption = null
        upload.setWidth("300px")

        // for mobile devices
        // upload.setImmediate(true);

        // progressbar update
        upload.addProgressListener { readBytes, contentLength -> bar.value = readBytes.toFloat() / contentLength.toFloat() }
        upload.addStartedListener {
            UI.getCurrent().pollInterval = 500
            progressLayout.isVisible = true
            upload.isVisible = false
            bar.value = 0f
            commentField.isVisible = false
            commentLabel.isVisible = false
            retentionCheckbox.isVisible = false
            retentionDateField.isVisible = false
            retentionDateLabel.isVisible = false
            filenameLabel.isVisible = false
            filenameField.isVisible = false
        }
        upload.addChangeListener { event ->
            val filename = event?.filename ?: ""
            val p = Paths.get(filename)
            filenameField.value = p.fileName.toString()
        }

        commentLabel.isVisible = true
        commentField.isVisible = true
        commentField.setWidth("250px")

        retentionCheckbox.isVisible = true
        retentionDateField.isVisible = false
        retentionDateLabel.isVisible = false
        retentionDateField.setWidth("250px")

        retentionDateField.resolution = Resolution.MINUTE

        retentionCheckbox.addValueChangeListener{
            retentionDateField.isVisible = retentionCheckbox.value
            retentionDateLabel.isVisible = retentionCheckbox.value
        }

        filenameLabel.isVisible = true
        filenameField.isVisible = true
        filenameField.setWidth("250px")

        content.addComponent(upload)
        content.addComponent(filenameLabel)
        content.addComponent(filenameField)
        content.addComponent(commentLabel)
        content.addComponent(commentField)
        content.addComponent(retentionCheckbox)
        content.addComponent(retentionDateLabel)
        content.addComponent(retentionDateField)

        uploadButton.setWidth("100px")
        uploadButton.addClickListener { upload.submitUpload() }

        content.addComponent(uploadButton)

    }

    companion object {
        private val logger = Logger.getLogger(UploadWindow::class.java.name)
    }

}
